### ISSUE TITLE
cant run ssh server properly inside a container

### ISSUE THREAD

**Issue by biocyberman** [apptainer/singularity#4319]
I first tried with NVIDIA's tensorflow image, then normal Ubuntu base image, and finally the SSH-enabled phusion's passenger-full image. All failed. So I would like to ask for help from singularity developers. If it helps anything, this is a email thread relating to this problem: https://groups.google.com/a/lbl.gov/d/topic/singularity/kjkPFsgd_t0/discussion

## Version of Singularity:

3.3.0

### Expected behavior
ssh server runs properly

### Actual behavior

can't login to the container. Running /usr/sbin/sshd -d reveals that i doesn't initialize authentication process:

```
/usr/sbin/sshd -d
debug1: sshd version OpenSSH_7.6, OpenSSL 1.0.2n  7 Dec 2017
debug1: private host key #0: ssh-rsa SHA256:EDWRBlyyVHfWKk/cAp2CI1GcnN1OLvxrSu8ay4jxQmM
debug1: private host key #1: ecdsa-sha2-nistp256 SHA256:mMeGj7V1XxyJSkmAKGqNsBYLqVq3C19n0fRkPnEGVpeM
debug1: private host key #2: ssh-ed25519 SHA256:lN7is7WPx95JBWUjcvj9GSsc44/dZ/X4xVTGJ5YL28M
debug1: rexec_argv[0]='/usr/sbin/sshd'
debug1: rexec_argv[1]='-d'
debug1: Set /proc/self/oom_score_adj from 0 to -1000
debug1: Bind to port 22 on 0.0.0.0.
Server listening on 0.0.0.0 port 22.
debug1: Bind to port 22 on ::.
Server listening on :: port 22.
debug1: Server will not fork when running in debugging mode.
debug1: rexec start in 5 out 5 newsock 5 pipe -1 sock 8
debug1: inetd sockets after dupping: 3, 3
Connection from 172.19.8.14 port 51288 on 10.23.0.22 port 22 <========================= Login attempt
debug1: Client protocol version 2.0; client software version OpenSSH_7.2p2 Ubuntu-4ubuntu2.8
debug1: match: OpenSSH_7.2p2 Ubuntu-4ubuntu2.8 pat OpenSSH* compat 0x04000000
debug1: Local version string SSH-2.0-OpenSSH_7.6p1 Ubuntu-4ubuntu0.3
debug1: permanently_set_uid: 105/65534 [preauth]
permanently_set_uid: was able to restore old [e]gid [preauth]
debug1: do_cleanup [preauth]
debug1: monitor_read_log: child log fd closed 
debug1: do_cleanup
debug1: Killing privsep child 50
debug1: audit_event: unhandled event 1
```

### Steps to reproduce behavior

#### Definition file (sshd.def):
```
Bootstrap: docker
From: phusion/passenger-full:1.0.6
# From: ubuntu:18.04
#From: nvidia/tensorflow:19.05-py3
#Registry: nvcr.io
IncludeCmd: yes

%environment
    export LANG=en_US.UTF-8

%post
    LANG=en_US.UTF-8
    echo $LANG UTF-8 > /etc/locale.gen
    apt-get install -y locales && update-locale --reset LANG=$LANG
    rm -f /etc/service/sshd/down
    # Regenerate SSH host keys. Passenger-docker does not contain any, so you
    # have to do that yourself. You may also comment out this instruction; the
    # init system will auto-generate one during boot.
    /etc/my_init.d/00_regen_ssh_host_keys.sh

%startscript
    /sbin/my_init --enable-insecure-key
```
#### Build

`singularity build --fakeroot sshd.sif sshd.def`

#### Run
```
singularity instance start -f --net --network-args="portmap=2222:22/tcp" ./sshd.sif sshd
INFO:    Convert SIF file to sandbox...
WARNING: underlay of /etc/localtime required more than 50 (129) bind mounts
*** Installing insecure SSH key for user root
*** Running /etc/my_init.d/00_regen_ssh_host_keys.sh...
*** Running /etc/my_init.d/10_syslog-ng.init...

INFO:    instance started successfully

singularity shell instance://sshd
Singularity rootfs-938503987:/tmp> service ssh status
 * sshd is not running
Singularity rootfs-938503987:/tmp> service ssh start
 * Starting OpenBSD Secure Shell server sshd                                                                                                                                                                                                                                                                                                                                                                            [ OK ]
Singularity rootfs-938503987:/tmp> service ssh status
 * sshd is running
Singularity rootfs-938503987:tmp>
```
ssh -p 2222 -l  root localhost
Connection reset by 127.0.0.1 port 2222

**Comment by jmstover** [comment]
Try running the instance with:  `--keep-privs` 

Also, there's an option `--boot` which you may find useful.

Also, are you using user namespace, or anything like that? 
```
debug1: permanently_set_uid: 105/65534 [preauth]
permanently_set_uid: was able to restore old [e]gid [preauth]
```
OpenSSH has some strict requirements on what it needs to be able to do. Dropbear is honestly the easiest thing to get working if you want to be able to SSH in.

**Comment by Flamefire** [comment]
I tried the same on a centos image running on a RHEL system and failed to get it to work. `strace`ing the sshd showed some DBUS failures and other stuff that I failed to resolve all so I gave up. Similar issues with Docker pop up in Google so this seems to be a general problem with running SSH inside a container.

> Also, there's an option --boot which you may find useful.

What does it really do for this issue? It runs init but does that help here?

> Also, are you using user namespace, or anything like that?

An instance automatically uses namespaces for e.g. PIDs and this can't be changed.

**Comment by biocyberman** [comment]
@Flamefire Thanks for answering the questions that I would give the same information. Regarding docker, the [pushion](https://github.com/phusion/baseimage-docker) folks actually say they can make ssh server work inside Docker. 

> What does it really do for this issue? It runs init but does that help here?

It would actually help if singularity can run init properly, then all service has a bootstrap to run on top.

I will give Dropbear a try if I still really need ssh. For now it seems that tuneling to singularity host and then do `singularity shell ...` would be sufficient for me. 

Nevertheless I do hope ssh works properly inside singularity so I can avoid hacky solutions.

**Comment by j-guitart** [comment]
This issue still persists with Singularity v3.4.2. Oddly enough, sshd works well with Singularity v3.1.0 running the same image, thus the cause should come from some change made between v3.1.0 and v.3.3.0.

**Comment by Dario-Mantegazza** [comment]
Any news on the issue?

**Comment by dtrudg** [comment]
For us to look into this further it'd be helpful to have the following:

- Output of '/usr/sbin/sshd -d' in a run on a container / singularity version that is working.
- Same output '/usr/sbin/sshd -d'  where it is not working

The output where it *is* working will help distinguish between things that aren't a critical issue and those that are.

Generally running ssh in a container is to be avoided where possible. Singularity has a different security model than docker, and I'm not surprised it doesn't work by default (as ssh will try to run with privilege separation etc.), but there should be some combination of options we can suggest for this.

**Comment by Dario-Mantegazza** [comment]
This is the result on the singularity where ssh is not working
```singularity shell --cleanenv --contain --net --network-args="portmap=2222:22/tcp" --fakeroot --bind ~/Desktop/mydata:/data --bind nomoko-idsia:/nomoko-idsia --bind nomoko-idsia-ml:/nomoko-idsia-ml nomoko-isin-1.5.sif
INFO:    Convert SIF file to sandbox...
Singularity nomoko-isin-1.5.sif:~> service ssh status
 * sshd is not running
Singularity nomoko-isin-1.5.sif:~> service ssh start
 * Starting OpenBSD Secure Shell server sshd                                                                                                                                                                [ OK ] 
Singularity nomoko-isin-1.5.sif:~> service ssh status
 * sshd is running
Singularity nomoko-isin-1.5.sif:~> /usr/sbin/sshd -d
debug1: sshd version OpenSSH_7.6, OpenSSL 1.0.2n  7 Dec 2017
debug1: private host key #0: ssh-rsa SHA256:c1X70ItHYmPwpu8ZVoWHwg6wiLFEV9Xeg0DUukiDRNY
debug1: private host key #1: ecdsa-sha2-nistp256 SHA256:TSVwl9rf0911CuB+lnm7S9q2VS/pKsnipn2UoWtM+/g
debug1: private host key #2: ssh-ed25519 SHA256:Bl16PDMuq91aLhOa5f1N3SMDqN5+uDMjoXKNPPxW0h0
debug1: rexec_argv[0]='/usr/sbin/sshd'
debug1: rexec_argv[1]='-d'
debug1: Set /proc/self/oom_score_adj from 0 to -1000
debug1: Bind to port 22 on 0.0.0.0.
Bind to port 22 on 0.0.0.0 failed: Address already in use.
debug1: Bind to port 22 on ::.
Bind to port 22 on :: failed: Address already in use.
Cannot bind any address.
Singularity nomoko-isin-1.5.sif:~> 
```
Unfortunately I do not have a container that is running ssh to test `/usr/sbin/sshd -d` there.
i'm using singularity version 3.4.0-1

**Comment by jmstover** [comment]
@Dario-Mantegazza 
Are you running in a new network namespace (container running as root)? The error you're getting is:

> Bind to port 22 on :: failed: Address already in use.

So ... you have something listening on port 22 already. I'm guessing the host SSH daemon. 

I always have used dropbear in the container running on a port >1024.

**Comment by j-guitart** [comment]
Hi, I provide the output of '/usr/sbin/sshd -d' where it is working vs. where it is not.

Singularity v3.1 (WORKING)

`debug1: sshd version OpenSSH_7.4, OpenSSL 1.0.2l  25 May 2017`
`debug1: private host key #0: ssh-rsa SHA256:3bIAPIkMphaeu58ovc8BruKVlUyBckkeA05TmZV8ymQ`
`debug1: private host key #1: ecdsa-sha2-nistp256 SHA256:jNcmgcupdzVCgsVvDP7wlQYSOpF4lq+DQQLJyReAIgs`
`debug1: private host key #2: ssh-ed25519 SHA256:deWN0HOWTU/lHtq09q80gtvl1ADMu0wgSVecKiSvUU4`
`debug1: rexec_argv[0]='/usr/sbin/sshd'`
`debug1: rexec_argv[1]='-d'`
`debug1: Set /proc/self/oom_score_adj from 0 to -1000`
`debug1: Bind to port 22 on 0.0.0.0.`
`Server listening on 0.0.0.0 port 22.`
`debug1: Bind to port 22 on ::.`
`Server listening on :: port 22.`
`debug1: Server will not fork when running in debugging mode.`
`debug1: rexec start in 5 out 5 newsock 5 pipe -1 sock 8`
`debug1: inetd sockets after dupping: 3, 3`
`Connection from 192.168.128.18 port 58124 on 192.168.128.2 port 22`
`debug1: Client protocol version 2.0; client software version OpenSSH_7.4p1 Debian-10+deb9u3`
`debug1: match: OpenSSH_7.4p1 Debian-10+deb9u3 pat OpenSSH* compat 0x04000000`
`debug1: Local version string SSH-2.0-OpenSSH_7.4p1 Debian-10+deb9u3`
`debug1: Enabling compatibility mode for protocol 2.0`
`debug1: permanently_set_uid: 102/65534 [preauth]`
`debug1: list_hostkey_types: ssh-rsa,rsa-sha2-512,rsa-sha2-256,ecdsa-sha2-nistp256,ssh-ed25519 [preauth]`
`debug1: SSH2_MSG_KEXINIT sent [preauth]`
`debug1: SSH2_MSG_KEXINIT received [preauth]`
`debug1: kex: algorithm: curve25519-sha256 [preauth]`
`debug1: kex: host key algorithm: ecdsa-sha2-nistp256 [preauth]`
`debug1: kex: client->server cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none [preauth]`
`debug1: kex: server->client cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none [preauth]`
`debug1: expecting SSH2_MSG_KEX_ECDH_INIT [preauth]`
`debug1: rekey after 134217728 blocks [preauth]`
`debug1: SSH2_MSG_NEWKEYS sent [preauth]`
`debug1: expecting SSH2_MSG_NEWKEYS [preauth]`
`debug1: SSH2_MSG_NEWKEYS received [preauth]`
`debug1: rekey after 134217728 blocks [preauth]`
`debug1: KEX done [preauth]`
`debug1: userauth-request for user root service ssh-connection method none [preauth]`
`debug1: attempt 0 failures 0 [preauth]`
`debug1: PAM: initializing for "root"`
`debug1: PAM: setting PAM_RHOST to "192.168.128.18"`
`debug1: PAM: setting PAM_TTY to "ssh"`
`debug1: userauth-request for user root service ssh-connection method publickey [preauth]`
`debug1: attempt 1 failures 0 [preauth]`
`debug1: userauth_pubkey: test whether pkalg/pkblob are acceptable for RSA SHA256:h6hxrnUyqG9UzKxHrBpsOTNi2M3B7mq6CkPoXbAB0FE [preauth]`
`debug1: temporarily_use_uid: 0/0 (e=0/0)`
`debug1: trying public key file /root/.ssh/authorized_keys`
`debug1: fd 4 clearing O_NONBLOCK`
`debug1: matching key found: file /root/.ssh/authorized_keys, line 1 RSA SHA256:h6hxrnUyqG9UzKxHrBpsOTNi2M3B7mq6CkPoXbAB0FE`
`debug1: restore_uid: 0/0`
`Postponed publickey for root from 192.168.128.18 port 58124 ssh2 [preauth]`
`debug1: userauth-request for user root service ssh-connection method publickey [preauth]`
`debug1: attempt 2 failures 0 [preauth]`
`debug1: temporarily_use_uid: 0/0 (e=0/0)`
`debug1: trying public key file /root/.ssh/authorized_keys`
`debug1: fd 4 clearing O_NONBLOCK`
`debug1: matching key found: file /root/.ssh/authorized_keys, line 1 RSA SHA256:h6hxrnUyqG9UzKxHrBpsOTNi2M3B7mq6CkPoXbAB0FE`
`debug1: restore_uid: 0/0`
`debug1: do_pam_account: called`
`Accepted publickey for root from 192.168.128.18 port 58124 ssh2: RSA SHA256:h6hxrnUyqG9UzKxHrBpsOTNi2M3B7mq6CkPoXbAB0FE`
`debug1: monitor_child_preauth: root has been authenticated by privileged process`
`debug1: monitor_read_log: child log fd closed`
`debug1: PAM: establishing credentials`
`debug1: rekey after 134217728 blocks`
`debug1: rekey after 134217728 blocks`
`debug1: ssh_packet_set_postauth: called`
`debug1: Entering interactive session for SSH2.`
`debug1: server_init_dispatch`
`debug1: server_input_channel_open: ctype session rchan 0 win 1048576 max 16384`
`debug1: input_session_request`
`debug1: channel 0: new [server-session]`
`debug1: session_new: session 0`
`debug1: session_open: channel 0`
`debug1: session_open: session 0: link with channel 0`
`debug1: server_input_channel_open: confirm session`
`debug1: server_input_global_request: rtype no-more-sessions@openssh.com want_reply 0`
`debug1: server_input_channel_req: channel 0 request pty-req reply 1`
`debug1: session_by_channel: session 0 channel 0`
`debug1: session_input_channel_req: session 0 req pty-req`
`debug1: Allocating pty.`
`debug1: session_pty_req: session 0 alloc /dev/pts/1`
`debug1: SELinux support disabled`
`debug1: server_input_channel_req: channel 0 request env reply 0`
`debug1: session_by_channel: session 0 channel 0`
`debug1: session_input_channel_req: session 0 req env`
`debug1: server_input_channel_req: channel 0 request env reply 0`
`debug1: session_by_channel: session 0 channel 0`
`debug1: session_input_channel_req: session 0 req env`
`debug1: server_input_channel_req: channel 0 request env reply 0`
`debug1: session_by_channel: session 0 channel 0`
`debug1: session_input_channel_req: session 0 req env`
`debug1: server_input_channel_req: channel 0 request env reply 0`
`debug1: session_by_channel: session 0 channel 0`
`debug1: session_input_channel_req: session 0 req env`
`debug1: server_input_channel_req: channel 0 request env reply 0`
`debug1: session_by_channel: session 0 channel 0`
`debug1: session_input_channel_req: session 0 req env`
`debug1: server_input_channel_req: channel 0 request env reply 0`
`debug1: session_by_channel: session 0 channel 0`
`debug1: session_input_channel_req: session 0 req env`
`debug1: server_input_channel_req: channel 0 request env reply 0`
`debug1: session_by_channel: session 0 channel 0`
`debug1: session_input_channel_req: session 0 req env`
`debug1: server_input_channel_req: channel 0 request env reply 0`
`debug1: session_by_channel: session 0 channel 0`
`debug1: session_input_channel_req: session 0 req env`
`debug1: server_input_channel_req: channel 0 request env reply 0`
`debug1: session_by_channel: session 0 channel 0`
`debug1: session_input_channel_req: session 0 req env`
`debug1: server_input_channel_req: channel 0 request env reply 0`
`debug1: session_by_channel: session 0 channel 0`
`debug1: session_input_channel_req: session 0 req env`
`debug1: server_input_channel_req: channel 0 request shell reply 1`
`debug1: session_by_channel: session 0 channel 0`
`debug1: session_input_channel_req: session 0 req shell`
`Starting session: shell on pts/1 for root from 192.168.128.18 port 58124 id 0`
`debug1: Setting controlling tty using TIOCSCTTY.`

Singularity v3.4.2 (NOT WORKING)

`debug1: sshd version OpenSSH_7.4, OpenSSL 1.0.2l  25 May 2017`
`debug1: private host key #0: ssh-rsa SHA256:3bIAPIkMphaeu58ovc8BruKVlUyBckkeA05TmZV8ymQ`
`debug1: private host key #1: ecdsa-sha2-nistp256 SHA256:jNcmgcupdzVCgsVvDP7wlQYSOpF4lq+DQQLJyReAIgs`
`debug1: private host key #2: ssh-ed25519 SHA256:deWN0HOWTU/lHtq09q80gtvl1ADMu0wgSVecKiSvUU4`
`debug1: rexec_argv[0]='/usr/sbin/sshd'`
`debug1: rexec_argv[1]='-d'`
`debug1: Set /proc/self/oom_score_adj from 0 to -1000`
`debug1: Bind to port 22 on 0.0.0.0.`
`Server listening on 0.0.0.0 port 22.`
`debug1: Bind to port 22 on ::.`
`Server listening on :: port 22.`
`debug1: Server will not fork when running in debugging mode.`
`debug1: rexec start in 5 out 5 newsock 5 pipe -1 sock 8`
`debug1: inetd sockets after dupping: 3, 3`
`Connection from 192.168.128.18 port 58270 on 192.168.128.2 port 22`
`debug1: Client protocol version 2.0; client software version OpenSSH_7.4p1 Debian-10+deb9u3`
`debug1: match: OpenSSH_7.4p1 Debian-10+deb9u3 pat OpenSSH* compat 0x04000000`
`debug1: Local version string SSH-2.0-OpenSSH_7.4p1 Debian-10+deb9u3`
`debug1: Enabling compatibility mode for protocol 2.0`
`debug1: permanently_set_uid: 102/65534 [preauth]`
`permanently_set_uid: was able to restore old [e]gid [preauth]`
`debug1: do_cleanup [preauth]`
`debug1: do_cleanup`
`debug1: Killing privsep child 113`
`debug1: audit_event: unhandled event 12`

**Comment by jmstover** [comment]
This is the error ... 
```
debug1: permanently_set_uid: 102/65534 [preauth]
permanently_set_uid: was able to restore old [e]gid [preauth]
```
@cclerget was there changes to UID/GID handling in Network namespace between 3.1.x and 3.4.2? Or is it touching user namespace which could be getting handled differently?

**Comment by cclerget** [comment]
@jmstover Here the problem :

https://github.com/sylabs/singularity/blob/16936f47dea8f46642517030af4e20a9adafe794/cmd/starter/c/starter.c#L231

It causes container processes to keep capabilities when setting UID/GID that's why sshd can change GID to non zero and switch back to 0 without loosing CAP_SETGID capability.

Fixed by  #4810

**Comment by jmstover** [comment]
@cclerget Thx.
@j-guitart Can you test PR #4810 ?

**Comment by j-guitart** [comment]
SSH works correctly with PR #4810 

Thanks for your work on this.

**Comment by jmstover** [comment]
@j-guitart Thanks much. We'll get it merged in here soon.

**Comment by biocyberman** [comment]
Singularity version 3.5.1

Some improvement, but still can't login to the server:

```
# Build sshd.sif as metioned previously
singularity instance start -f --net --network-args="portmap=2222:22/tcp" --writable ./sshd.sif sshd
singularity shell instance://sshd
ingularity> passwd
Enter new UNIX password:
Retype new UNIX password:
passwd: password updated successfully
Singularity> /etc/init.d/ssh restart
 * Restarting OpenBSD Secure Shell server sshd                                                                                                                                [ OK ]
Singularity> passwd # Set password for root
Enter new UNIX password:
Retype new UNIX password:
passwd: password updated successfully
Singularity> vim /etc/ssh/sshd_config #Enable root login
Singularity> /etc/init.d/ssh restart
 * Restarting OpenBSD Secure Shell server sshd                                                                                                                                [ OK ]

# Logging into the ssh from another terminal
 ssh -p 2222 -l root localhost
The authenticity of host '[localhost]:2222 ([127.0.0.1]:2222)' can't be established.
ECDSA key fingerprint is SHA256:dfqnXzs/WaWT2LIMS3UtvlWLDbtFusAV1KLSyzg0sBQ.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added '[localhost]:2222' (ECDSA) to the list of known hosts.
root@localhost's password:
Connection to localhost closed by remote host.
Connection to localhost closed.

```

Doing a debug session showed the below log:

```
debug1: Bind to port 22 on 0.0.0.0.
Server listening on 0.0.0.0 port 22.
debug1: Bind to port 22 on ::.
Server listening on :: port 22.
debug1: Server will not fork when running in debugging mode.
debug1: rexec start in 5 out 5 newsock 5 pipe -1 sock 8
debug1: inetd sockets after dupping: 3, 3
Connection from 10.23.0.1 port 42264 on 10.23.0.35 port 22
debug1: Client protocol version 2.0; client software version OpenSSH_7.2p2 Ubuntu-4ubuntu2.8
debug1: match: OpenSSH_7.2p2 Ubuntu-4ubuntu2.8 pat OpenSSH* compat 0x04000000
debug1: Local version string SSH-2.0-OpenSSH_7.6p1 Ubuntu-4
debug1: permanently_set_uid: 101/65534 [preauth]
debug1: list_hostkey_types: ssh-rsa,rsa-sha2-512,rsa-sha2-256,ecdsa-sha2-nistp256,ssh-ed25519 [preauth]
debug1: SSH2_MSG_KEXINIT sent [preauth]
debug1: SSH2_MSG_KEXINIT received [preauth]
debug1: kex: algorithm: curve25519-sha256@libssh.org [preauth]
debug1: kex: host key algorithm: ecdsa-sha2-nistp256 [preauth]
debug1: kex: client->server cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none [preauth]
debug1: kex: server->client cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none [preauth]
debug1: expecting SSH2_MSG_KEX_ECDH_INIT [preauth]
debug1: rekey after 134217728 blocks [preauth]
debug1: SSH2_MSG_NEWKEYS sent [preauth]
debug1: expecting SSH2_MSG_NEWKEYS [preauth]
debug1: SSH2_MSG_NEWKEYS received [preauth]
debug1: rekey after 134217728 blocks [preauth]
debug1: KEX done [preauth]
debug1: userauth-request for user root service ssh-connection method none [preauth]
debug1: attempt 0 failures 0 [preauth]
debug1: userauth-request for user root service ssh-connection method password [preauth]
debug1: attempt 1 failures 0 [preauth]
Accepted password for root from 10.23.0.1 port 42264 ssh2
debug1: monitor_child_preauth: root has been authenticated by privileged process
debug1: monitor_read_log: child log fd closed
debug1: rekey after 134217728 blocks
debug1: rekey after 134217728 blocks
debug1: ssh_packet_set_postauth: called
debug1: Entering interactive session for SSH2.
debug1: server_init_dispatch
debug1: server_input_channel_open: ctype session rchan 0 win 1048576 max 16384
debug1: input_session_request
debug1: channel 0: new [server-session]
debug1: session_new: session 0
debug1: session_open: channel 0
debug1: session_open: session 0: link with channel 0
debug1: server_input_channel_open: confirm session
debug1: server_input_global_request: rtype no-more-sessions@openssh.com want_reply 0
debug1: server_input_channel_req: channel 0 request pty-req reply 1
debug1: session_by_channel: session 0 channel 0
debug1: session_input_channel_req: session 0 req pty-req
debug1: Allocating pty.
debug1: session_pty_req: session 0 alloc /dev/pts/98
debug1: SELinux support disabled
chown(/dev/pts/98, 0, 5) failed: Operation not permitted
debug1: do_cleanup
debug1: session_pty_cleanup: session 0 release /dev/pts/98
debug1: audit_event: unhandled event 12
Singularity> exit
```

`chown(/dev/pts/98, 0, 5) failed: Operation not permitted` seems to be noteworthy.

**Comment by cclerget** [comment]
@biocyberman As you use fakeroot and you didn't specify `--contain` flag, `/dev/pts` in container is the one from the host and of course you don't have right to change ownership.
You have two options in this case :
- use `--contain` flag to automatically mount a new instance of devpts
- in the startscript add `mount -t devpts devpts /dev/pts` before starting ssh server

**Comment by biocyberman** [comment]
> @biocyberman As you use fakeroot and you didn't specify `--contain` flag, `/dev/pts` in container is the one from the host and of course you don't have right to change ownership.
> You have two options in this case :
> 
>     * use `--contain` flag to automatically mount a new instance of devpts
> 
>     * in the startscript add `mount -t devpts devpts /dev/pts` before starting ssh server

That's it. Thanks. I infer from what you said, that --fakeroot is optional? How do I run sshd without fakeroot? And how do I actually start the ssh service, login to the container running ssh with normal user, not to mention AD user?

**Comment by cclerget** [comment]
> --fakeroot is optional ? How do I run sshd without fakeroot?

Either fakeroot or start instance as root user since ssh requires to change identity.

> And how do I actually start the ssh service

Please look at the documentation : https://sylabs.io/guides/3.5/user-guide/definition_files.html#startscript, to start you will need to add a `startscript` section in your definition file:

```
%startscript
    service ssh start
```

> login to the container running ssh with normal user, not to mention AD user?

If you want to manage identity inside your container, you need to configure your container and start services accordingly like you would do on your server.
